CM/CD: Add polygon source

tonynajjar opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | N/A |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Custom gazebo simulation |

---

## Description of contribution in a few bullet points

We would like to add a source to the collision monitor and that is an array of arbitrary PolygonStamped fed from a topic. Our specific use case is that we would like to prevent collisions with some tracked objects. 

Note: Adding these tracked objects to the global/local costmap was tried, however this was not enough as the controller makes a "best effort" to avoid the collisions; we need more decisive logic to stop the robot and therefore need the collision monitor

* Add Polygon as a source
* Create new msg type PolygonArray to feed in polygons to the Polygon source


## Description of documentation updates required from your changes

<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

---

## Future work that may be required in bullet points

<!--
* I think there might be some optimizations to be made from STL vector
* I see alot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

tonynajjar commented on nav2_util/src/robot_utils.cpp:
I needed the output as geometry_msgs::msg::TransformStamped to keep the stamp. We can think of a less verbose solution maybe later

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
could maybe be refined to not reflect the whole message

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
result:

![image](https://github.com/ros-planning/navigation2/assets/14879660/ebb4bdd9-03e2-4e8e-8cdf-7825ef9b57f8)

tonynajjar commented:
@SteveMacenski @AlexeyMerzlyakov that's a first draft, a lot of small improvements can still be made but wanted your opinion on the idea before spending more time refining it. Thanks

SteveMacenski commented on nav2_collision_monitor/src/collision_monitor_node.cpp:
Why not just `polygon`? (param + file name + class name)

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
We do have a `LineIterator` that you may be able to use instead to make this easer. You can set the start/end coordinates and march along it by `increments`. We use it mostly in the `map` coordinate space so we march by "1" cell, but you can give it world coordinates and march by another increment if you prefer

SteveMacenski commented on nav2_util/src/robot_utils.cpp:
Could this be called by the nonstamped version above so we reduce duplicate code?

SteveMacenski reviewed (commented):
Test coverage?

I'd think for polygon-source types, we would want to make the minimum number of points 1 for this special case

AlexeyMerzlyakov commented on nav2_msgs/msg/PolygonsArray.msg:
Since each `PolygonStamped` in the array below has its own header, I think, main header is unnecessary

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
Since each polygon in the array might be in its own frame, I believe, we need to get transform per each polygon it the message instead of using main headers' frame_id

AlexeyMerzlyakov commented on nav2_util/src/robot_utils.cpp:
I'd also like to get rid of duplicate code, as @SteveMacenski mentioned by utilizing `tf2::Transform`. Another alternative could be making this routine as inline and calling it from inside non-stamped `getTransform` version

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
std::hypot can do this

tonynajjar commented on nav2_collision_monitor/src/collision_monitor_node.cpp:
Just made it the same as the filename/class but you're right, doesn't have to be, changing

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
hmm I gave it a quick look but it's not clear to me how to use it in this situation, the increment doesn't seem to be configurable. I removed an obsolete part of it, maybe now it's a bit simpler. Is it a must for you to use the LineIterator here?

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
good point, changed

tonynajjar commented on nav2_util/src/robot_utils.cpp:
I think I found a satisfying solution, I changed the output of `getTransform` from `tf2::Transform` to `tf2::Stamped<tf2::Transform>` so that we can keep the stamp information. The other sources are not affected since `    tf2::Vector3 p_v3_b = tf_transform * p_v3_s;` should work with both stamped and unstamped Transform

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
done

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
Any opinions on what to consider a valid message?

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
This should use standard messages, not custom ones. You can have it store a vector of data and then clear the vector after its been populated by the monitor

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
FYI @doisyg for timeout-ing: how should we handle if its multiple data? If any single one is bad or if all are bad?

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
Space above this line

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
No space

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
Space between functions

Overall, linting / styling here is weird. Please fix.

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
While this works, again, why not use the line iterator we have in nav2_utils?

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
I'm thinking we just only include the current ones and log a warning about any expired ones that aren't being used -- but not returning false if any are valid.

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
hmm, I see one challenge, how would you not store duplicates in the vector? By cross-checking every field of the msg::Polygon against the Polygons already in the vector and not adding it if it matches?

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
Will these polygons have headers? Perhaps keep the latest of each header typed? Maybe as a parameter so that others might want to use all

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
They will need to have headers to check individually if they are too old to be valid. What do yoh mean with keep it "typed", can you clarify?

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
I mean if you have N polygons, check the header values and keep only the newest of each reference frame, but as a parameterized option

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
What would you propose?

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
I stated the only alternatives I have already, either custom message or tracking with std::set + custom comparator as mentioned above. The custom msg seems simpler but I understand wanting to stick to standard messages...

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
OK, lets go with that then. I'm not sure a custom comparator is needed, I believe `msg == msg` is implemented for all ROS message structs.

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
Is not just `1` w/o static_cast won't be suitable here?

AlexeyMerzlyakov commented on nav2_collision_monitor/test/sources_test.cpp:
The question to make sure that everything is syched: the first point seems to asymmetrical towards to the rest 3. Is it the initial intention to make non-bar polygon, or misprint?

AlexeyMerzlyakov commented on nav2_collision_monitor/test/sources_test.cpp:
Also, the case to cover the working with the `PolygonSource` is needed in `collision_monitor_node_test.cpp` and `collision_detector_node_test.cpp`. It could be new simple testcase, just create the (1,1,-1,-1) stop polygon and add there polygon source with intersection with stop polygon, like depicted below, and make sure that robot will stop (for CM; or report stopping for CD) after polygon source will entered/interlaced the stop polygon:
![20231026_154946](https://github.com/ros-planning/navigation2/assets/60094858/4dcb14a2-50a1-4582-9139-b4aa416448d6)

AlexeyMerzlyakov reviewed (commented):
There is almost nothing to add to @SteveMacenski comments from last review on the main logic. Please check the remaining items for the testcases from below

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
I implemented the logic that will store incoming polygons to a vector plus some additional logic to keep this vector up to date:

- if a new polygon arrives that is equivalent to an existent one in the vector but is more recent, then it replaces it
- if any polygon is older than source_timeout it's cleared from the vector

The problem I'm facing is that if we're publishing a "moving polygon", then there will be a trail (of the old poses of this polygon) because the old poses in the vector do not satisfy any of the above stated conditions initially. Only after they are older than source_timeout will they be cleared

![image](https://github.com/ros-planning/navigation2/assets/14879660/ea68e40e-c74a-464a-a1ad-ecb40b664a04)

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
probably, GPT generated that part :p

tonynajjar commented on nav2_collision_monitor/test/sources_test.cpp:
good point, not intentional

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
nope, fails:

```
/code/ros2_ws/src/navigation2/nav2_collision_monitor/src/polygon_source.cpp:125:15: error: no matching function for call to ‘max(size_t, int)’
  125 |       std::max(static_cast<size_t>(segment_length / sampling_distance_), 1);
```

doisyg commented on nav2_collision_monitor/src/polygon_source.cpp:
I did not dig in this PR. I will probably do once https://github.com/ros-planning/navigation2/pull/3880 is in and if this one is rebased on it to check all timeout related behavior.

doisyg commented on nav2_collision_monitor/src/polygon_source.cpp:
> FYI @doisyg for timeout-ing: how should we handle if its multiple data? If any single one is bad or if all are bad?

Behavior wise, I think from the moment one source is bad (but with https://github.com/ros-planning/navigation2/pull/3880 we could have more complex behavior with independent per-source configuration)

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
It doesn't need to be size_t.

Also, please add a comment anywhere generative AI was used to generate code. I'm working on a proposal around that and it would be good practice for me to see if my suggestions are helpful :-)

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
That's what I thought about the frame_id containing some unique element of the particular polygon - but I could see how that wouldn't work if everything is pre-transformed. I'd really ideally like to **not** create a new collision-monitor specific message type, since that wouldn't be portable & it would require a shim to any natural messages they might use. A name or uuid would do the trick, but not having something like that in the msg definition causes a problem. 

I suppose another option is to have `N` polygon sources each of their own topic for each individual polygon. For these polygons you're using, how many are we talking about would be of interest? 1? 2?  10? 100? Would having 1 topic-source per robot footprint be doable?

Another option: if the polygons are stamped and we make some statement that the polygons to run multiple in 1 instance need to be sychronized, we could check for any that have a timestamp older than the latest by `eta` and remove those. Think something like a depth processing pipeline with stereo image pairs. Its a little hacky. 

The last option is the first you had; just have a custom message. I suppose we could do that, though I'd probably want that to include the polygon, stamp, and some unique identifier so we can fully solve the problem if we had that. The array might not be necessary at that point then to allow it to be more distributed. Though, I suppose it could continue to be a vector as long as we don't just store the last message but buffer `N` data sources

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
I implemented my proposal and it works quite well, I tried to minimize the risk of wrongfully clearing polygons by making `arePolygonsSimilar` check 2 things, the polygons are considered similar if:

1- they have the same number of points
2- Pair by pair, their respective points all have a distance smaller than `polygon_similarity_threshold_`

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
I ran ament_uncrustify on it, please specify what you consider weird so I can fix it.

tonynajjar commented on nav2_collision_monitor/test/sources_test.cpp:
fixed

tonynajjar commented on nav2_collision_monitor/src/collision_monitor_node.cpp:
follow up for https://github.com/foxglove/studio/issues/6991

tonynajjar commented:
I thin most points have been addressed, please have another look.

SteveMacenski commented:
@AlexeyMerzlyakov please re-review

@tonynajjar this file has by far the most missing coverage of the Collision Monitor files https://app.codecov.io/gh/ros-planning/navigation2/blob/add-polygon-source-main/nav2_collision_monitor%2Fsrc%2Fpolygon_source.cpp#L179 please fill in the gaps. I'm not worried about the `node.lock()` exceptions, but there is a complete function missing an a for loop totally unexecuted.

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
Missing doxygen

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
Missing a check on timeouts if this data is valid using the object constructor argument `source_timeout`

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
What part of it, specifically? Or the entire function?

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
I can see why you thought you could remove the check with source timeout from the main file -- but there's no promise that this callback is run in any particular frequency or regularity. We still need to check the polygon's timeout in the main getData function

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
Better yet, move this up there -- we're looping through the data anyway in `getData`, do the deletion there and remove here

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
@smac is the proposes solution satisying to you? (I'll worry abou the testing at the end to avoid double work)

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
or equal to, if they're literally the same?

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
How does this generalize?

SteveMacenski reviewed (changes requested):
Generally looks good, but I don't think the are polygons similar approach is valid general approach to this. This assumes alot about what specifically these polygons are, how often they're published, and that they're spatially near each other due to that regular frequency

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
I think if you want to replace data, we need to have (1) some kind of unique identifier or (2) have unique frames of each thing that we use TF to transform. (2) feels fragile, so I think the answer here is to have some unique identifier in the messages of the physical object the polygon belongs to. This feels like a good add to the standard messages too, let me inquire if we can make an API change

https://github.com/ros2/common_interfaces/issues/230

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
For the unique ID, we could abuse the 'seq' of the Header msg. I say "abuse" because it's more meant as a sequence counter.

Regarding your PR, you could also make the case to adding the unique identifier to the Header msg, since this uid could be needed for many other msgs, e.g. Point, Vector, etc...

(I'm on vacation for the whole week; I'll try to answer but it might take a while or not be as detailled as I would wish)

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
There is no seq in ROS 2 :smiling_face_with_tear: 

No worries, enjoy vacation. I'd be disappointed if I heard from you the rest of the week :P

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
Damn it, you're right I was looking at the ROS 1 version. Second part of my comment is still valid though.

SteveMacenski commented:
@tonynajjar when you're back, rebase and fix the API changes for `bool getData` and checking the timeout via https://github.com/ros-planning/navigation2/pull/3880 merging

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
For the case, if no Unique ID to be used: this could be done through `std::hypot(dx, dy, dz)`

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
How do you handle moving polygon challenge? For example, the polygon is moving in some frame and its coordinates are being updated&published e.g. 5 times per second. In this case, this routine will add N-polygons at the same time unless old ones won't exceed `source_timeout_`. Changing `source_timeout_` to be `0.2 sec` to meet 5-times per a second rate handling one polygon at a time, seems to be not a good idea as `Source::sourceValid()` checker is also using the same timeout to determine whether the data is rust. And robot in this case might go with some "jerking" by stopping-running-stopping...

AlexeyMerzlyakov commented on nav2_collision_monitor/src/polygon_source.cpp:
I'd more prefer the idea with polygon `UUID` that @SteveMacenski mentioned in previous messages.
... or using polygons array from previous implementation seem also doable idea for me.

AlexeyMerzlyakov reviewed (commented):
The main challenge here - is how to solve movin polygon task, where polygon changing its coordinates during the time.
Whether should we use `unique_identifier_msgs/msg/UUID.msg`, or get back to previous implementation when all polygons are being published once per a time, or might be something else (which I did not noted).

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
You're right that the removal of outdated polygons can't be done here; I had it in the getData function originally but the problem is that the getData function is marked as `const`, meaning I can't modify `data_` in it. Are you fine with removing the const from that function? Keep in mind it has to be removed for all sources.

tonynajjar commented on nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp:
done

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
all of it (checked by me of course), edited

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
discussed in comment below about getData being const

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
true, done

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
done

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
no, removed

tonynajjar commented:
Addressed most of the comments, I'll wait for the result of https://github.com/ros2/common_interfaces/issues/230 to continue this PR  and the tests. It's also my preferred solution to the problem

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
While I agree that an id in the header could be useful, I think that is a uphill battle, but is worth a discussion!

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
Yeah, I'm fine with that!

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
@tonynajjar I'm looking into this more for writing ROS community guidelines and it looks like this is illegal. You can't copy right chatgpt code and can't use it unless you found the source it pulled it from to check the appropriate license. You would need to have substantively modified the gpt-generated code for it to be legal.

Please reimplement yourself

tonynajjar commented on nav2_collision_monitor/src/polygon_source.cpp:
What if you asked GPT to generate a bubble sort algorithm for you, would that be illegal to use? Some things don't have multiple implementations... is changing the variable names + minor tweaks enough?

SteveMacenski commented on nav2_collision_monitor/src/polygon_source.cpp:
In my current understanding of the law: yes, it is illegal. You should reimplement without looking at this methods current implementation.

SteveMacenski commented:
@tonynajjar Polygons with IDs have been added! https://github.com/ros2/common_interfaces/pull/232#event-11529615113

That can be used now to implement this PR to get over the issue :-)

tonynajjar commented:
Awesome! In the meantime I left my old company that needed this and am quite busy with my new one. @jplapp maybe you want to take this over? If not I'll try to eventually find the time but can't promise when.

SteveMacenski commented:
It should be straight forward now if using those messages. All of the `arePolygonsSimilar` functionality can be deleted and just check the IDs. Add a few unit tests and a nice video and I think its done.

jplapp commented:
Thanks for review! Sure, I'll find someone to complete this, but cannot promise a timeline

oldtopos commented:
@jplapp I have cycles to pick this up; it seems there is one use of arePolygonsSimilar to replace, so this should be straightforward.  LMK if that works for you

oldtopos commented:
I am attempting to copy the branch associated with the PR to my own fork of nav2 in github, as I presume (and have not tested) I do not have commit access to the repo currently associated with the merge.  Will see how the build goes tomorrow, fingers crossed
